aws/logs_monitoring/cache.py

```diff
@@ -0,0 +1,421 @@
+        if self._is_expired():
+            send_forwarder_internal_metrics("local_cache_expired")
+            logger.debug("Local cache expired, fetching cache from S3")
+            self._refresh()
+
+        function_tags = self.tags_by_id.get(key, [])
+        return function_tags
+
+
+#############################
+# Cloudwatch Log Group Tags #
+#############################
+
+cloudwatch_logs_client = boto3.client("logs")
+
+
+def get_log_group_tags(log_group):
+    response = None
+    try:
+        send_forwarder_internal_metrics("list_tags_log_group_api_call")
+        response = cloudwatch_logs_client.list_tags_log_group(logGroupName=log_group)
+    except Exception as e:
+        logger.exception(f"Failed to get log group tags due to {e}")
+    if response is not None:
+        formatted_tags = [
+            "{key}:{value}".format(
+                key=sanitize_aws_tag_string(k, remove_colons=True),
+                value=sanitize_aws_tag_string(v, remove_leading_digits=False),
+            )
+            if v
+            else sanitize_aws_tag_string(k, remove_colons=True)
+            for k, v in response["tags"].items()
+        ]
+    else:
+        # In the case of failing to query tags we will write an empty list and retry on the next refresh
+        formatted_tags = []
+    return formatted_tags
+
+
+class CloudwatchLogGroupTagsCache(LambdaTagsCache):
+    CACHE_FILENAME = DD_S3_LOG_GROUP_CACHE_FILENAME
+    CACHE_LOCK_FILENAME = DD_S3_LOG_GROUP_CACHE_LOCK_FILENAME
+
+    def build_tags_cache(self):
+        """Makes API calls to GetResources to get the live tags of the account's Lambda functions
+
+        Returns an empty dict instead of fetching custom tags if the tag fetch env variable is not set to true
+
+        Returns:
+            tags_by_arn_cache (dict<str, str[]>): each Lambda's tags in a dict keyed by ARN
+        """
+        new_tags = {}
+        for log_group in self.tags_by_id.keys():
+            new_tags[log_group] = get_log_group_tags(log_group)
+
+        logger.debug("All tags in Cloudwatch Log Groups refresh: {}".format(new_tags))
+        return True, new_tags
+
+    def get(self, log_group):
+        """Get the tags for the Cloudwatch Log Group from the cache
+
+        Will refetch the tags if they are out of date, or a log group is encountered
+        which isn't in the tag list
+
+        Args:
+            key (str): the key we're getting tags from the cache for
+
+        Returns:
+            log_group_tags (str[]): the list of "key:value" Datadog tag strings
+        """
+        if self._is_expired():
+            send_forwarder_internal_metrics("local_cache_expired")
+            logger.debug("Local cache expired, fetching cache from S3")
+            self._refresh()
+
+        log_group_tags = self.tags_by_id.get(log_group, None)
+        if log_group_tags is None:
+            log_group_tags = get_log_group_tags(log_group)
+            self.tags_by_id[log_group] = log_group_tags
```

Comment:
Is this needed? I think we're updating `self.tags_by_id` in the `_refresh` function

Reply:
Right, we'll get here if we're encountering a log group that isn't in `self.tags_by_id`, so it'll be a log group we haven't gotten logs from before. In this case we grab tags for it and add it temporarily to  `self.tags_by_id`. It'll get added to the s3 cache when we refresh.